Deploy command created.

rhatlapa opened the pull request:
Created deploy command, which handles deploy part of #42.

**EDIT**: updated link to point to correct issue.

Ladicek commented:
How is #10 related? Should be #42, no?

Ladicek commented:
An option -- CLI has `deploy` which could be used here. I'm of course by no means objecting to creating the mgmt operation manually, just suggesting a possibility.

rhatlapa commented:
Yes, you are right, I meant #42

rhatlapa commented:
I wanted to use management operation directly to prevent need of creating temporary files and to just leave it in memory.

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/deployments/DeployCmdTest.java:
```diff
@@ -0,0 +1,88 @@
+/**
+ * This test uses the {@code ajp} socket binding for all listener types, because there's no listener configured
+ * by default that uses it (unlike {@code http}).
+ */
```
Does it? :-)

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,206 @@
+    private final boolean autoCloseInputStream;
+    private final List<String> serverGroups;
+
+    private Deploy(InputStream deploymentInputStream, Data deployCmdOptions) {
```
This is slightly worrying, because `InputStream` is a one-off thing. If we were on Java 8, I'd prefer `Supplier<InputStream>`. Another option would be Guava's `ByteSource`, and even though we already depend on Guava, I'd like to avoid depending on it in public APIs. Another very similar thing is ShrinkWrap's `Asset`. Sadly, on Java 6, there's probably nothing we can do (short of providing our own interface, which I'd really like to avoid) :-(

rhatlapa replied:
Agree with you, don't now about any better way with Java 6 than using the InputStream. If that would be big issue, I would redo this and allow only File. In that case only in memory handling would be lost :-(.

Ladicek replied:
No need to do that, we can live with `InputStream` for now.

Ladicek commented:
What are your plans with this? Do you want this reviewed and merged, or do you plan to work on undeploy commands first and prefer to keep this PR open?

rhatlapa commented:
I plan to work on undeploy command, I believe I should have time to do it during this week. If no one is in hurry, I would wait for it :-).

rhatlapa commented:
I have added also the undeploy command, now it fixes #42 as whole :-).

Note I was thinking also about removing all options from undeploy command, that it would in case of domain always undeploy the deployment from all relevant server groups. As the way the undeploy command behaves, it requires using enabling option for keeping content if there wasn't all relevant server groups selected. WDYT?

rhatlapa commented:
I've decided to remove the option for specifying server groups as I believe no one really wants to use it this way. If it is found needed, I left it as builder => it should be easy to add those options without breaking API.

rhatlapa commented:
=> ready for review

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,211 @@
+    /**
+     * A single shared instance of this class is passed through the entire builder invocation chain to accumulate
+     * all the options.
+     */
```
This is clearly copied from `Online|OfflineOptions`, where I deliberately broke the builder to a bunch of classes to be able to provide compile-time checking of all mandatory options. It adds quite some maintenance burden, so I don't think it's suitable for commands. Did you have any special reason to use this pattern instead of plain old single-class builder?

rhatlapa replied:
I just wanted to have also check at compile time, but I agree it is not so important in this case and having written in javadoc which options are only for domain should be satisfactory.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,211 @@
+                groupDeploymentAddresses.add(mn);
+            }
+            deployRequest = createDeploymentOperation(content,
+                    groupDeploymentAddresses.toArray(new ModelNode[groupDeploymentAddresses.size()]));
```
I wonder if `Operations.batch` would work here. Maybe not, in which case we should figure out whether the usecases in this command are rare or common, and possibly expand the `Operations` API a bit.

rhatlapa replied:
I don't believe it is possible to use `Operations.batch` here. Partial help would be to expose `batch.toModelNode`. As I would be able to create the needed `ModelNode` for constructing `org.jboss.as.controller.client.Operation` which is needed here.

Ladicek replied:
Aha, the `Operation` thing. So far, I'm only aware of a single thing that needs to attach streams to operations: deployments. Let's stick with `ModelNode`s here for now.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,211 @@
+            this.autoCloseInputStream = autoCloseInputStream;
+        }
+
+        public Data(File deploymentFile) {
```
The constructors shouldn't be `public`.

Ladicek replied:
In fact, probably nothing in this class should be `public`.

rhatlapa replied:
You are right, good point, as I will redo it to standard builder the whole Data class should disappear.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,211 @@
+            this.deployCmdData = new Data(deploymentFile);
+        }
+
+        public DeployCmdBuilder(final InputStream deploymentInputStream, final String deploymentName,
```
We typically don't mark parameters `final`, unless they have to be.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,211 @@
+        /**
+         * Adds list of server groups for which the deployment should be applied.
+         */
+        public Deploy toParticularServerGroups(String... serverGroups) {
```
I think `toServerGroups` would be shorter and better (more symmetric with `toAllServerGroups`).

rhatlapa replied:
good point, will change the name

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/deployments/DeployUndeployCmdTest.java:
```diff
@@ -0,0 +1,118 @@
+/**
+ * This class tests deploy and undeploy commands in standalone mode.
+ */
+@Category(WildFlyTests.class)
```
Is there a particular reason why this test doesn't work with AS7/EAP6? I think we certainly want these commands to work there.

rhatlapa replied:
I just stupidly copied the annotation without really thinking about it. It should work also with AS7/EAP6.

Ladicek replied:
Good! We want this command tested for all server versions for sure.

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/deployments/DeployUndeployCmdTest.java:
```diff
@@ -0,0 +1,118 @@
+public class DeployUndeployCmdTest {
+
+    private static final String DEPLOYMENT_NAME = "test-deployment.war";
+    private static final Address TEST_DEPLOYMENT_ADDRESS = Address.of("deployment", DEPLOYMENT_NAME);
```
I think we should add `Address.deployment`.

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/deployments/DeployUndeployCmdTest.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private void deployAsFile(Archive archive) throws CommandFailedException, IOException, OperationException {
+        File testDeploymentFile = new File(tmp.getRoot(), archive.getName());
```
`tmp.newFile(archive.getName())`

rhatlapa replied:
that would actually create the new file, I don't want that. The export in next step shall create it.

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/deployments/DeployUndeployCmdTest.java:
```diff
@@ -0,0 +1,118 @@
+
+    private void assertDeploymentExists(String deploymentName, boolean shouldExist)
+            throws IOException, OperationException {
+        Assert.assertEquals("Deployment exists", shouldExist, ops.exists(Address.of("deployment", deploymentName)));
```
The assert message only makes sense for `shouldExist == true`.

Ladicek replied:
I mean:

```java
String assertMessage = shouldExist ? "Deployment should exist" : "Deployment shouldn't exist";
```

rhatlapa replied:
I believe that the message would look like: "Deployments exists, expected ...., actual ....". Maybe I've got confused by how the messages are created with TestNG vs JUnit. If it would be as I expect, I believe my original message should be fine. Will check it.

Ladicek replied:
For Creaper, I prefer a "should" style of assertion messages. If you expect the deployment to exist, the assertion message would read "deployment should exist"; if you expect the deployment to not exist, the assertion message would read "deployment shouldn't exist".

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Undeploy.java:
```diff
@@ -0,0 +1,56 @@
+        if (ctx.options.isDomain) {
+            cmd.append(" --all-relevant-server-groups");
+        }
+        ctx.client.executeCli(cmd.toString());
```
Is there a reason you use the "local" CLI operation `undeploy` here, but for `deploy`, you assemble the `ModelNode` by hand?

rhatlapa replied:
The reason is that the "local" CLI operation handles finding server groups with which the deployment is associated and I haven't found any helper method in the Helper classes related to `ModelControllerClient` to help me find them => I would have to duplicate code used by the CLI.

`deploy` directly via ModelNode allows to do the deployment just via in memory storing of the deployment, via CLI I would have to save the deployment as file.

Ladicek replied:
OK then.

rhatlapa commented:
BTW: there is also option to rewrite the `deploy` command and make use of `DeploymentPlan` class similarly as arquillian is handling deploying.

Ladicek commented:
If you feel like using `DeploymentPlan` would make the implementation easier, go for it :-) I'm fine with it both ways.

rhatlapa commented:
I have incorporated your suggestions, should be ready for another round of review :-)

Ladicek commented:
LGTM. Please squash.

I'd also like @simkam and @ochaloup to check if the API exposes all the features they need.

Ladicek commented:
Also please rebase, I just pushed #61 :-) Shouldn't affect this PR, you're not doing anything version-specific.

I'd also like to do a release today, so if you don't rebase & squash soon, this will have to wait for another release. Waiting for review from @simkam and @ochaloup is not a prerequisite for merge, though.

ochaloup commented:
For me it's find. I haven't found anything special that I would need for more from the API.

simkam commented on commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java:
```diff
@@ -0,0 +1,176 @@
+            if (serverGroups != null) {
+                this.serverGroups.addAll(Arrays.asList(serverGroups));
+            }
+            return build();
```
I talked to Radim and he mentioned that he is not sure if this ok. I agree that this could problem in future if we eventually add more options to builder. @Ladicek wdyt?

Ladicek replied:
It's also a bit unusual. I agree it would be better if these methods just returned `Builder` and the user always had to call `build()`.

Ladicek replied:
On the other hand, as Radim just told me, it makes sure that the user can't call both `toServerGroups` and `toAllServerGroups`, which is nice. IMHO it would be enough to throw an exception when it's detected that the user called both of these methods, the question is how to detect this -- that would require slightly more complex encoding in the `Builder`.

rhatlapa commented:
I have done force update, where I have squashed the commits, rebased and modified the builder to return the builder object for `toServerGroups` and `toAllServerGroups` and emphasize in javadoc that only one of those should be used, not both of them.

Ladicek commented:
OK, thanks.
